### ISSUE TITLE
Programmatically added listeners from ServletContextListener.contextInitialzed() are not called

### ISSUE THREAD

**Issue by olamy** [eclipse/jetty.project#3799]
Not tested but I suppose the issue is for all EventListener.
Pseudo code in a ServletContextListener
```
            AddListServletRequestListener listenerClass =
                sce.getServletContext().createListener( AddListServletRequestListener.class );
            sce.getServletContext().addListener( listenerClass );
```
The ServletRequestListener is not executed for any requests

**Comment by olamy** [comment]
Test included in branch ` jetty-9.4.x_issue_3799_add_servlet_request_listener `
Run it using 
`  mvn clean -T3 install -Dinvoker.test=FOO -Dtest=DynamicListenerTests -pl :test-distribution -am  `

**Comment by janbartel** [comment]
The Spec says it is illegal to add a ServletContextListener from a ServletContextListener, see section 4.4.3.2.  The only place it is legal to programmatically add a ServletContextListener is from a ServletContainerInitializer. We have (manual)tests for this, see the tests/test-webapps/test-servlet-spec/test-spec-webapp/src/main/java/com/acme/test/TestListener class, and tests/test-webapps/test-servlet-spec/test-container-initializer/src/main/java/com/acme/initializer/FooInitializer.java

**Comment by olamy** [comment]
The failing test is adding a ServletRequestListener from a ServletContextListener.
tck test https://github.com/eclipse-ee4j/jakartaee-tck/blob/master/src/com/sun/ts/tests/servlet/api/javax_servlet/servletcontext30/TestListener.java#L167
the test I added https://github.com/eclipse/jetty.project/commit/219eb51341fd95badfbb742bec74b9f5211d0d47

**Comment by janbartel** [comment]
@olamy seems like this could be a bug with the way listeners are added from a ServletContextListener.contenxtInitialized method. I'll look into it.

**Comment by janbartel** [comment]
@olamy got a fix, working on comprehensive test case, stay tuned for a PR.

**Comment by janbartel** [comment]
Fixed for 9.4.20 and above.
